_posts/2021-08-16-connector-table-sql-api-part2.md

```diff
@@ -0,0 +1,497 @@
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (folder != null) {
+            folder.close();
+        }
+
+        if (store != null) {
+            store.close();
+        }
+    }
+}
+```
+
+There is a request trigger to the server in every loop iteration. This is crucial as it ensures that the server will keep sending notifications. A more sophisticated approach would be to make use of the IDLE protocol. 
+
+Note that since the source is not checkpointable, no state fault tolerance will be possible.
+
+
+## Collect incoming emails
+
+Now you need to listen for new emails arriving in the inbox folder and collect them. To begin, hardcode the schema and only return the email’s subject. Fortunately, Jakarta provides a simple hook (`addMessageCountListener()`) to get notified when new messages arrive on the server. You can use this in place of the “TODO” comment above:
+
+```java
+import jakarta.mail.*;
+import jakarta.mail.event.MessageCountAdapter;
+import jakarta.mail.event.MessageCountEvent;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+
+public class ImapSource extends RichSourceFunction<RowData> {
+    @Override
+    public void run(SourceFunction.SourceContext<RowData> ctx) throws Exception {
+        // …
+
+        folder.addMessageCountListener(new MessageCountAdapter() {
+            @Override
+            public void messagesAdded(MessageCountEvent e) {
+                collectMessages(ctx, e.getMessages());
+            }
+        });
+
+        // …
+    }
+
+    private void collectMessages(SourceFunction.SourceContext<RowData> ctx, Message[] messages) {
+        for (Message message : messages) {
+            try {
+                ctx.collect(GenericRowData.of(StringData.fromString(message.getSubject())));
+            } catch (MessagingException ignored) {}
+        }
+    }
+}
+```
+
+Now build the project again and start up the SQL client:
+
+```sh
+$ cd testing/
+$ ./build_and_run.sh
+```
+
+This time, you will connect to a [GreenMail server](https://greenmail-mail-test.github.io/greenmail/) which is started as part of the [setup](https://github.com/TNG/flink-connector-email/blob/master/testing/docker-compose.yaml):
+
+```sql
+CREATE TABLE T (
+    subject STRING
+) WITH (
+    'connector' = 'imap', 
+    'host' = 'greenmail',
+    'port' = '3143', 
+    'user' = 'alice', 
+    'password' = 'alice'
+);
+
+SELECT * FROM T;
+```
+
+The query above should now run continuously but no rows will be produced since it is a test server. You need to first send an email to the server. If you have [mailx](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/mailx.html) installed, you can do so by executing in your terminal:
+
+```sh
+$ echo "This is the email body" | mailx -Sv15-compat \
+        -s"Email Subject" \
+        -Smta=imap: \
+        alice@acme.org
+
```

Comment:
I'm not sure where this command came from, but it doesn't work. Here's what I had in my Google Doc and verified again to work:

```suggestion
$ echo "This is the email body" | mailx -Sv15-compat \
        -s"Email Subject" \
        -Smta="smtp://alice:alice@localhost:3025" \
        alice@acme.org
```